File: web_version/pmugg_files/graphController.js
// The controller for the graph family tree.
ms.graphController = function(view, mainController) {
	this.view = view;
	this.mainController = mainController;
	this.viewport = null;
	this.cameraController = new ms.cameraController(view, this.notify.bind(this));
	this.classifier = new ms.classifier();
	this.familyTree = [];
};

ms.graphController.actionButtons = {
	CLASSIFY: 0,
};

ms.graphController.prototype.notify = function() {
	ms.timerG.start('Redraw');
	var expectedMode = this.classifier.is3D ? ms.view.modes.VIEW3D : ms.view.modes.SHAPE;
	if (this.view.mode != expectedMode) {
		this.view.setMode(expectedMode);
		this.viewport = this.view.getViewport();
		this.cameraController.activate();
	}
	this.view.redraw(this);
	ms.timerG.stop('Redraw');
};

ms.graphController.prototype.activate = function(opt_mode, opt_callback) {
	if (this.mainController.getExample().isEmpty()) {
		this.mainController.addTestObject();
	}
	this.view.setMode(ms.view.modes.SHAPE, ms.mainController.controllerTypes.GRAPH);
	this.mainController.setModeButtons([]);
	this.mainController.setActionButtons(ms.mvp ? [] : ['Classify']);
	this.viewport = this.view.getViewport();
	this.viewport.setBaseScale(0);
	this.cameraController.activate();
	ms.globalMutator = this;
	this.classify();
};

ms.graphController.prototype.applyAction = function(button) {
	var actionButtons = ms.graphController.actionButtons;
	if (button == actionButtons.CLASSIFY) {
		this.classify();
	}
};

ms.graphController.prototype.classify = function() {
	ms.edgeType.count = 0;
	ms.graph.count = 0;
	ms.graphEdge.count = 0;
	ms.counter.reset();
	var example = this.mainController.getExample();
	example.updateFaces();
	
	if (example.solution) {
		this.classifier.importSolution(example.solution);
		this.classifier.is3D = example.shapes[0].is3D;
	} else {		
		ms.timerG.start('Create Grammar');
		this.classifier.setShapes(example.getShapes());
		ms.timerG.stop('Create Grammar');
	}
	
	this.notify();
	console.log(ms.timerG.reportCurrent());
};

ms.graphController.prototype.getRenderables = function() {
	return [this.classifier.getFamilyTree()];
};

ms.graphController.prototype.offset = function() {
	return new ms.vec2(0, 0);
};

ms.graphController.prototype.onmousemove = function (event) {
	this.cameraController.onmousemove(event);
};

ms.graphController.prototype.onmouseup = function (event) {
	this.cameraController.onmouseup(event);
};

ms.graphController.prototype.onmousedown = function (event) {
	this.cameraController.onmousedown(event);
};


ms.graphController.prototype.onmouseout = function (event) {};

ms.graphController.prototype.onDoubleClick = function (event) {};

ms.graphController.prototype.onRightClick = function (event) {};

ms.graphController.prototype.onEscape = function (event) {};

ms.graphController.prototype.onKeyPress = function (event) {
	var keycode = event.keyCode;
	switch (keycode) {
		default:	break;
	}
};

ms.graphController.prototype.onKeyUp = function (event) {};

File: cpp_version/pmugg/graph_controller.cpp
#include "graph_controller.h"
#include "timer.h"

namespace ms {

GraphController::GraphController(View* view, MainController* mainController)
    : view(view)
    , mainController(mainController)
    , viewport(nullptr)
    , cameraController(new CameraController(view, [this]() { this->notify(); }))
    , classifier(new Classifier())
    , familyTree() {
}

void GraphController::notify() {
    Timer::start("Redraw");
    
    auto expectedMode = classifier->is3D ? View::Mode::VIEW3D : View::Mode::SHAPE;
    if (view->getMode() != expectedMode) {
        view->setMode(expectedMode);
        viewport = view->getViewport();
        cameraController->activate();
    }
    
    view->redraw(this);
    Timer::stop("Redraw");
}

void GraphController::activate(int mode, std::function<void()> callback) {
    if (mainController->getExample()->isEmpty()) {
        mainController->addTestObject();
    }

    std::vector<std::string> actionButtons = {"Classify"};
    mainController->setModeButtons({});
    mainController->setActionButtons(actionButtons);

    view->setMode(View::Mode::SHAPE);
    viewport = view->getViewport();
    cameraController->activate();

    if (callback) {
        callback();
    }

    notify();
}

void GraphController::classify() {
    Timer::start("Classify");
    classifier->classify(mainController->getExample());
    Timer::stop("Classify");
    notify();
}

void GraphController::onMouseMove(const MouseEvent& event) {
    cameraController->onMouseMove(event);
}

void GraphController::onMouseDown(const MouseEvent& event) {
    cameraController->onMouseDown(event);
}

void GraphController::onMouseUp(const MouseEvent& event) {
    cameraController->onMouseUp(event);
}

void GraphController::onMouseOut(const MouseEvent& event) {
    // Handle mouse out event if needed
}

void GraphController::onDoubleClick(const MouseEvent& event) {
    // Handle double click event if needed
}

void GraphController::onRightClick(const MouseEvent& event) {
    // Handle right click event if needed
}

void GraphController::onEscape(const KeyEvent& event) {
    // Handle escape key event if needed
}

void GraphController::onKeyPress(const KeyEvent& event) {
    switch (event.keyCode) {
        default:
            break;
    }
}

void GraphController::onKeyUp(const KeyEvent& event) {
    // Handle key up events if needed
}

} // namespace ms 